mitodl/mit-open
# Add nesting learning resource topics

### What are the relevant tickets?

Closes mitodl/hq#4089

### Description (What does it do?)

This PR allows topics for learning resources to be nested, so we can have a hierarchy of topics. This involves:
- Some model changes to allow for nesting
- A management command to bulk load the topic data from OCW (or compatible source)
- Minor updates to the API to support parents

The importer expects to pull these from the OCW course site config. This is available [here](https://github.com/mitodl/ocw-studio/blob/dd823c6eec258650508b3498ff5d8657738ba9ea/static/js/resources/ocw-course-site-config.json). Use the `update_topics` management command to kick off upserting items from the file. The system expects the OCW JSON file to live in `learning_resources/data/` but you can also specify a path if you wish. 

You can technically also add topics from other sources if they're in the same format OCW stores them in. 

When storing a LearningResource record, it will queue a task to automatically backfill parent topics, so if you select a topic that's nested under another topic, the resource will have all the applicable topics. (Note that this is queued so if you change this in, say, Django Admin, you will need to reload the record to see the parent topics be added.) 

When backfilling/importing data from other data sources, any topics that are listed in the data source but that don't exist in Open will be skipped. A warning message gets emitted to the log when this happens. (This is a change from before - it would create the topics instead.)

Topics should otherwise work as expected, though.

### How can this be tested?

Automated tests should pass.

Test importing using the `update_topics` command. It should work - you will need to get an OCW course site config file (linked above).

Test adding a learning resource to a topic: 
- Test loading from the ETL pipelines - run the backfill commands or queue tasks. These should process successfully and things that have topics should have them attached accordingly. No new topics should be created.
- Manual testing - edit a learning resource in Django Admin, and add a topic or two to some. For best results, including testing search and channels, add topics that are nested - i.e., add "Engineering" to one resource, then add something like "Biomaterials" to another. The second should get "Biomaterials", plus "Biological Engineering" and "Engineering" as topics. 

Test creating a channel for a topic: Create a field channel record of type "Topic" with search filter `topic=<a topic>`, then update the search index. Loading the channel URL should show you the resources that have that topic assigned. If you use the same scenario listed above, create a channel for "Engineering" and one for "Biomaterials"; you should see two courses in the first and one in the second. 

Test searching: searching for a topic should work as expected. (Channels work using search so you should get back the same stuff in a search that you'd get back for an equivalent channel.)
